### PR TITLE
change(rust): Use link-time optimisation in release builds

### PR DESCRIPTION
**Pull request by teor2345**
## Motivation

As part of #4155, we want to speed up the `zebrad` initial sync.

Rust can optimise code at link-time, which reduces binary size, and seems to increase sync speed by about 5 minutes.

### Specifications

https://doc.rust-lang.org/cargo/reference/profiles.html#lto

## Solution

- Turn on link-time optimisation for Rust code in release builds

Related cleanups:
- remove redundant optimisation flag from the Dockerfile
    - [optimisation level 3 is already the default for release](https://doc.rust-lang.org/book/ch14-01-release-profiles.html#customizing-builds-with-release-profiles)

This makes the `zebrad` binary about 50 MB smaller on my machine, and the sync about 5 minutes faster.

## Review

Anyone can review this PR. It makes the sync faster, so it's a high priority.

### Reviewer Checklist

  - [ ] Code implements Specs and Designs
  - [ ] Tests for Expected Behaviour
  - [ ] Tests for Errors

### CODE REVIEW

**Comment by teor2345** [comment]
I'm running a full sync test here:
https://github.com/ZcashFoundation/zebra/actions/runs/2218514256

This PR should be faster than the checkpoint PR.

File: Cargo.toml [hunk 1]
```diff
@@ -51,6 +51,14 @@ opt-level = 1
 [profile.release]
 panic = "abort"
 
+# Speed up release syncs and sync tests by optimizing all release build code.
+# Some of Zebra's code is CPU-intensive, and needs extra optimizations for peak performance.
+#
+# The `cc` crate automatically applies this setting when compiling C/C++ code:
+# https://docs.rs/cc/latest/cc/struct.Build.html#method.opt_level
+# https://doc.rust-lang.org/cargo/reference/environment-variables.html#environment-variables-cargo-sets-for-build-scripts
+opt-level = 3
```
Comment (conradoplg) [inline Cargo.toml]: [This is already the default for `release`](https://doc.rust-lang.org/book/ch14-01-release-profiles.html#customizing-builds-with-release-profiles)

**Review by conradoplg — COMMENTED** [review]
This didn't seem to improve the sync time since opt-level=3 is already the default for release.

The other changes seem unrelated, did you mean to include them?

**Comment by conradoplg** [comment]
> The other changes seem unrelated, did you mean to include them?

Ah nevermind, this was caused by the automatic rebase

**Comment by teor2345** [comment]
> This PR should be faster than the checkpoint PR.

It's about 5 minutes faster.

File: Cargo.toml (hunk 1 above)
Comment (teor2345) [reply Cargo.toml]: Hmm, that's weird, I guess the 5 minutes was network variation.

I'll try turning on link-time optimisation instead.

**Comment by teor2345** [comment]
I'm running a full sync test for LTO here:
- https://github.com/ZcashFoundation/zebra/actions/runs/2222788739

We should merge if it is the same speed or faster than #4183, which was 5h53m.

**Comment by teor2345** [comment]
> I'm running a full sync test for LTO here:
> 
>     * https://github.com/ZcashFoundation/zebra/actions/runs/2222788739
> 
> 
> We should merge if it is the same speed or faster than #4183, which was 5h53m.

Failed due to #3991, restarted at:
https://github.com/ZcashFoundation/zebra/actions/runs/2222788739

**Comment by teor2345** [comment]
> We should merge if it is the same speed or faster than #4183, which was 5h53m.

Failed due to #3991, cherry-picked fix #4198, and restarted at:
https://github.com/ZcashFoundation/zebra/actions/runs/2225002822

**Comment by conradoplg** [comment]
> > We should merge if it is the same speed or faster than #4183, which was 5h53m.
> 
> Failed due to #3991, cherry-picked fix #4198, and restarted at: https://github.com/ZcashFoundation/zebra/actions/runs/2225002822

Worked in 5h47m :tada: 

but failed macos-latest test for some reason (["Error: The operation was canceled."](https://github.com/ZcashFoundation/zebra/runs/6170895026?check_suite_focus=true), but it didn't timeout?). I'll retry it

**Review by conradoplg — APPROVED** [review]
Looks good, but I'm assuming the retry logic will be reviewed in #4198

**Comment by teor2345** [comment]
macOS failed with:
> Message:  Opening metrics endpoint listener 127.0.0.1:55555 failed: Hyper(hyper::Error(Listen, Os { code: 48, kind: AddrInUse, message: "Address already in use" })). Hint: Check if another zebrad or zcashd process is running. Try changing the metrics endpoint_addr in the Zebra config.

https://github.com/ZcashFoundation/zebra/runs/6178143289?check_suite_focus=true#step:13:1439

Port `55555` was chosen randomly, but maybe it's in use on GitHub runners. (It seems suspiciously regular.)

**Comment by teor2345** [comment]
@mergifyio update

**Comment by teor2345** [comment]
Mergify queue PR #4184 failed with:
```
#27 exporting to image
#27 pushing layers 0.5s done
#27 ERROR: failed to authorize: failed to fetch oauth token: unexpected status: 401 Unauthorized
------
 > importing cache manifest from us-docker.pkg.dev/zealous-zebra/zebra/zebrad-test:4210-merge-buildcache:
------
------
 > exporting to image:
------
error: failed to solve: failed to fetch oauth token: unexpected status: 401 Unauthorized
Error: buildx failed with: error: failed to solve: failed to fetch oauth token: unexpected status: 401 Unauthorized
```
https://github.com/ZcashFoundation/zebra/runs/6185722377?check_suite_focus=true#step:9:2180

**Comment by teor2345** [comment]
@mergifyio refresh

**Pull request merged**